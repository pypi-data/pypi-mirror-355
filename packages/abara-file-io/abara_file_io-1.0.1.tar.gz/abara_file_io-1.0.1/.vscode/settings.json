{
  "[python]": {
    "files.encoding": "utf8",
    "files.eol": "\n",
    "files.autoGuessEncoding": false,
    "editor.tabSize": 4,
    "editor.insertSpaces": true,
    "editor.detectIndentation": false,
    "editor.defaultFormatter": "charliermarsh.ruff",
    "editor.formatOnSave": true,
    "editor.codeActionsOnSave": {
      "source.fixAll": "explicit",
      "source.organizeImports": "explicit",
    },
  },
  "ruff.configurationPreference": "filesystemFirst",
  "python.testing.pytestArgs": [
    "tests",
    "--cov=src",
  ],
  "python.testing.unittestEnabled": false,
  "python.testing.pytestEnabled": true
}
