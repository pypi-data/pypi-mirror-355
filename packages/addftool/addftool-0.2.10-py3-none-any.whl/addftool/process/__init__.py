import psutil
import time

from .utils import get_processes, get_process_using_rocm, get_process_using_cuda


def add_killer_args(subparsers):
    process_killer_parser = subparsers.add_parser('kill', help='process kill')

    process_killer_parser.add_argument("-c", "--contains", help="contains of command", action='store_true', default=False)
    process_killer_parser.add_argument("--timeout", help="timeout of command", default=5, type=int)
    process_killer_parser.add_argument("--try_count", help="try count of command", default=3, type=int)

    process_killer_parser.add_argument("--contain_arg", help="args of command", default="", type=str)

    process_killer_parser.add_argument("--rocm", help="kill process using rocm", action='store_true', default=False)
    process_killer_parser.add_argument("--cuda", help="kill process using cuda", action='store_true', default=False)

    process_killer_parser.add_argument("-v", "--view", help="view process", action='store_true', default=False)

    process_killer_parser.add_argument("name", nargs='?', help="name of process", type=str, default="")


def kill_process(processes, timeout=5, try_count=3):
    for process in processes:
        try:
            proc = psutil.Process(process['pid'])
            print(f"Killing process {proc.pid} - {proc.cmdline()}")
            proc.terminate()
        except (psutil.NoSuchProcess, psutil.AccessDenied, psutil.ZombieProcess):
            print(f"Process {process['pid']} not found or access denied")
            continue
    
    time.sleep(timeout)
    for _ in range(try_count):
        all_terminated = True
        for process in processes:
            try:
                proc = psutil.Process(process['pid'])
                print("Process still running: ", proc.pid)
                proc.kill()
                all_terminated = False
            except (psutil.NoSuchProcess):
                continue
            except (psutil.AccessDenied):
                print(f"Process {process['pid']} access denied")
                continue
            except (psutil.ZombieProcess):
                print(f"Process {process['pid']} is a zombie process")
                continue
            except Exception as e:
                print(f"Error killing process {process['pid']}: {e}")
                continue
        if all_terminated:
            print("All processes terminated")
            break
        time.sleep(timeout)
    
    if not all_terminated:
        print("Some processes are still running. Please check manually.")
        exit(1)


def find_and_kill_process(command="", contains=False, contain_arg="", use_rocm=False, use_cuda=False, timeout=5, try_count=3, only_view=False):
    do_not_do_anything = command is None or len(command) == 0
    if use_rocm:
        processes = get_process_using_rocm()
    elif use_cuda:
        processes = get_process_using_cuda()
    elif do_not_do_anything and not only_view:
        print("Use top or htop to find the process you want to kill")
        return
    else:
        processes = get_processes(command=command, contains=contains, contain_arg=contain_arg)
    
    if only_view:
        print(f"Found {len(processes)} processes")
        for process in processes:
            print(f"PID: {process['pid']}, Command: {' '.join(process['command'])}")
        return

    if len(processes) > 0:
        print(f"Found {len(processes)} processes to kill")
        kill_process(processes, timeout, try_count)


def killer_main(args):
    print(args)
    find_and_kill_process(
        args.name, args.contains, args.contain_arg, use_rocm=args.rocm, use_cuda=args.cuda, 
        timeout=args.timeout, try_count=args.try_count, only_view=args.view,
    )
