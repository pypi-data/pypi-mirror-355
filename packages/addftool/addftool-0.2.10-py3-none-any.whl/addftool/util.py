import os
import subprocess


def execute_command(command, to_file=None, only_stdout=True, hide=False):
    if to_file is not None:
        to_file.write(command + "\n")
        return None
    else:
        if not hide:
            print("Execute command: ", command)
        result = subprocess.Popen(command, shell=True, stdout=subprocess.PIPE)
        result.wait()
        print(f"Return code: {result.returncode}")
        if result.stdout is not None:
            stdout = result.stdout.read().decode()
            print(f"Stdout: {stdout}")
        else:
            stdout = None
        if only_stdout:
            if not hide and stdout is not None:
                print(stdout)
            return stdout
        if result.stderr is not None:
            stderr = result.stderr.read().decode()
            print(f"Stderr: {stderr}")
        else:
            stderr = None

        return {'stdout': stdout, 'stderr': stderr, 'returncode': result.returncode}


def need_sudo():
    return os.name == 'posix' and os.getuid() != 0


def is_running_in_docker():
    return os.path.exists('/.dockerenv') or \
           any('docker' in line for line in open('/proc/self/cgroup', 'r')) if os.path.exists('/proc/self/cgroup') else False or \
           os.environ.get('container') == 'docker' or \
           os.environ.get('DOCKER') == 'true' or \
           os.environ.get('DOCKER_CONTAINER') == 'yes'


def get_ubuntu_version():  
    with open("/etc/os-release") as f:
        for line in f:
            if line.startswith("VERSION_ID="):
                version = line.split("=")[1].strip().strip('"')
                return version
    return "22.04"


def check_package_installed(package):
    command = f"dpkg -l | grep {package}"
    result = execute_command(command)
    if result is not None and package in result:
        return True
    return False


def install_packages(package_list):
    to_install = []
    for package in package_list:
        if check_package_installed(package):
            print(f"{package} is already installed")
            continue
        to_install.append(package)
    
    if len(to_install) > 0:
        packages = " ".join(to_install)
        command = f"apt-get install -y {packages}"
        if need_sudo():
            command = "sudo " + command
        execute_command(command)
