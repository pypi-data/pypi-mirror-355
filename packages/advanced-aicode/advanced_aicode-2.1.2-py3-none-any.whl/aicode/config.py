from dataclasses import asdict, dataclass, field, fields
from typing import Dict, Optional, Union

from aicode.aider_update_result import AiderUpdateResult


@dataclass
class Config:
    openai_key: Optional[str] = None
    anthropic_key: Optional[str] = None
    gemini_key: Optional[str] = None
    aider_update_info: Dict[str, Union[str, bool, None]] = field(default_factory=dict)

    @property
    def aider_update_result(self) -> AiderUpdateResult | None:
        if self.aider_update_info:
            return AiderUpdateResult.from_json(self.aider_update_info)
        return None

    @staticmethod
    def from_dict(data: dict) -> "Config":
        # Only extract keys that match the dataclass fields
        valid_keys = {f.name for f in fields(Config)}
        filtered_data = {k: v for k, v in data.items() if k in valid_keys}
        return Config(**filtered_data)

    @staticmethod
    def load() -> "Config":
        from aicode.openaicfg import load_from_storage

        data: dict = load_from_storage()
        return Config.from_dict(data)

    def to_dict(self) -> dict:
        return asdict(self)

    def save(self) -> None:
        from aicode.openaicfg import save_config

        save_config(self.to_dict())
