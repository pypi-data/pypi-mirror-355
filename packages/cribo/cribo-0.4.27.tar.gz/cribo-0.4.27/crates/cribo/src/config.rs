use anyhow::{Context, Result, anyhow};
use indexmap::IndexSet;
use serde::{Deserialize, Serialize};
use std::env;
use std::path::{Path, PathBuf};

use crate::combine::Combine;
use crate::dirs::{system_config_file, user_cribo_config_dir};

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(default)]
pub struct Config {
    /// Source directories to scan for first-party modules
    pub src: Vec<PathBuf>,

    /// Known first-party module names
    pub known_first_party: IndexSet<String>,

    /// Known third-party module names
    pub known_third_party: IndexSet<String>,

    /// Whether to preserve comments in output
    pub preserve_comments: bool,

    /// Whether to preserve type hints in output
    pub preserve_type_hints: bool,

    /// Target Python version for standard library and builtin checks
    /// Supports Ruff-style string values: "py38", "py39", "py310", "py311", "py312", "py313"
    /// Defaults to "py310" (Python 3.10)
    #[serde(rename = "target-version")]
    pub target_version: String,
}

impl Default for Config {
    fn default() -> Self {
        Self {
            src: vec![PathBuf::from("src"), PathBuf::from(".")],
            known_first_party: IndexSet::new(),
            known_third_party: IndexSet::new(),
            preserve_comments: true,
            preserve_type_hints: true,
            target_version: "py310".to_owned(),
        }
    }
}

impl Combine for Config {
    fn combine(self, other: Self) -> Self {
        Self {
            // For collections, higher precedence (self) completely replaces lower precedence (other)
            // if self has non-default values, otherwise use other
            src: if self.src != Config::default().src {
                self.src
            } else {
                other.src
            },
            known_first_party: if !self.known_first_party.is_empty() {
                self.known_first_party
            } else {
                other.known_first_party
            },
            known_third_party: if !self.known_third_party.is_empty() {
                self.known_third_party
            } else {
                other.known_third_party
            },
            // For scalars, self always takes precedence
            preserve_comments: self.preserve_comments,
            preserve_type_hints: self.preserve_type_hints,
            target_version: self.target_version,
        }
    }
}

/// Configuration values from environment variables with CRIBO_ prefix
#[derive(Debug, Clone, Default)]
pub struct EnvConfig {
    pub src: Option<Vec<PathBuf>>,
    pub known_first_party: Option<IndexSet<String>>,
    pub known_third_party: Option<IndexSet<String>>,
    pub preserve_comments: Option<bool>,
    pub preserve_type_hints: Option<bool>,
    pub target_version: Option<String>,
}

impl EnvConfig {
    /// Load configuration from environment variables with CRIBO_ prefix
    pub fn from_env() -> Self {
        let mut config = Self::default();

        // CRIBO_SRC - comma-separated list of source directories
        if let Ok(src_str) = env::var("CRIBO_SRC") {
            let paths: Vec<PathBuf> = src_str
                .split(',')
                .map(|s| s.trim())
                .filter(|s| !s.is_empty())
                .map(PathBuf::from)
                .collect();
            if !paths.is_empty() {
                config.src = Some(paths);
            }
        }

        // CRIBO_KNOWN_FIRST_PARTY - comma-separated list of first-party modules
        if let Ok(first_party_str) = env::var("CRIBO_KNOWN_FIRST_PARTY") {
            let modules: IndexSet<String> = first_party_str
                .split(',')
                .map(|s| s.trim())
                .filter(|s| !s.is_empty())
                .map(|s| s.to_owned())
                .collect();
            if !modules.is_empty() {
                config.known_first_party = Some(modules);
            }
        }

        // CRIBO_KNOWN_THIRD_PARTY - comma-separated list of third-party modules
        if let Ok(third_party_str) = env::var("CRIBO_KNOWN_THIRD_PARTY") {
            let modules: IndexSet<String> = third_party_str
                .split(',')
                .map(|s| s.trim())
                .filter(|s| !s.is_empty())
                .map(|s| s.to_owned())
                .collect();
            if !modules.is_empty() {
                config.known_third_party = Some(modules);
            }
        }

        // CRIBO_PRESERVE_COMMENTS - boolean flag
        if let Ok(preserve_comments_str) = env::var("CRIBO_PRESERVE_COMMENTS") {
            config.preserve_comments = parse_bool(&preserve_comments_str);
        }

        // CRIBO_PRESERVE_TYPE_HINTS - boolean flag
        if let Ok(preserve_type_hints_str) = env::var("CRIBO_PRESERVE_TYPE_HINTS") {
            config.preserve_type_hints = parse_bool(&preserve_type_hints_str);
        }

        // CRIBO_TARGET_VERSION - target Python version
        if let Ok(target_version) = env::var("CRIBO_TARGET_VERSION") {
            config.target_version = Some(target_version);
        }

        config
    }

    /// Apply environment config to base config
    pub fn apply_to(self, mut config: Config) -> Config {
        if let Some(src) = self.src {
            config.src = src;
        }
        if let Some(known_first_party) = self.known_first_party {
            config.known_first_party = known_first_party;
        }
        if let Some(known_third_party) = self.known_third_party {
            config.known_third_party = known_third_party;
        }
        if let Some(preserve_comments) = self.preserve_comments {
            config.preserve_comments = preserve_comments;
        }
        if let Some(preserve_type_hints) = self.preserve_type_hints {
            config.preserve_type_hints = preserve_type_hints;
        }
        if let Some(target_version) = self.target_version {
            config.target_version = target_version;
        }
        config
    }
}

/// Parse a boolean value from string, supporting various common formats
fn parse_bool(value: &str) -> Option<bool> {
    use cow_utils::CowUtils;
    match value.cow_to_lowercase().as_ref() {
        "true" | "1" | "yes" | "on" => Some(true),
        "false" | "0" | "no" | "off" => Some(false),
        _ => None,
    }
}

impl Config {
    /// Parse a Ruff-style target version string to u8 version number
    /// Supports: "py38" -> 8, "py39" -> 9, "py310" -> 10, "py311" -> 11, "py312" -> 12, "py313" -> 13
    pub fn parse_target_version(version_str: &str) -> Result<u8> {
        match version_str {
            "py38" => Ok(8),
            "py39" => Ok(9),
            "py310" => Ok(10),
            "py311" => Ok(11),
            "py312" => Ok(12),
            "py313" => Ok(13),
            _ => Err(anyhow!(
                "Invalid target version '{}'. Supported versions: py38, py39, py310, py311, py312, py313",
                version_str
            )),
        }
    }

    /// Get the Python version as u8 for compatibility with existing code
    pub fn python_version(&self) -> Result<u8> {
        Self::parse_target_version(&self.target_version)
    }

    /// Set the target version from a string value
    pub fn set_target_version(&mut self, version: String) -> Result<()> {
        // Validate the version string
        Self::parse_target_version(&version)?;
        self.target_version = version;
        Ok(())
    }

    /// Load a single config file from a path
    pub fn load_from_file<P: AsRef<Path>>(path: P) -> Result<Config> {
        let path = path.as_ref();
        let content = std::fs::read_to_string(path)
            .with_context(|| format!("Failed to read config file: {:?}", path))?;

        let config: Config = toml::from_str(&content)
            .with_context(|| format!("Failed to parse config file: {:?}", path))?;

        // Validate the target version
        config.python_version().with_context(|| {
            format!(
                "Invalid target-version in config file: {}",
                config.target_version
            )
        })?;

        Ok(config)
    }

    fn try_load_and_combine<P: AsRef<Path>>(
        config: &mut Config,
        path: P,
        context: &str,
    ) -> Result<()> {
        if path.as_ref().exists() {
            log::debug!("Loading {} from: {:?}", context, path.as_ref());
            let loaded = Self::load_from_file(&path)
                .with_context(|| format!("Failed to load {} from {:?}", context, path.as_ref()))?;
            *config = loaded.combine(config.clone());
        }
        Ok(())
    }

    /// Load configuration with hierarchical precedence:
    /// 1. CLI-provided config path (highest precedence)
    /// 2. Environment variables (CRIBO_*)
    /// 3. Project config (cribo.toml in current directory)
    /// 4. User config (~/.config/cribo/cribo.toml)
    /// 5. System config (/etc/cribo/cribo.toml or equivalent)
    /// 6. Default values (lowest precedence)
    pub fn load(cli_config_path: Option<&Path>) -> Result<Self> {
        let mut config = Config::default();

        // 1. Load system config (lowest precedence)
        if let Some(system_config_path) = system_config_file() {
            Self::try_load_and_combine(&mut config, &system_config_path, "system config")?;
        }

        // 2. Load user config
        if let Some(user_config_dir) = user_cribo_config_dir() {
            let user_config_path = user_config_dir.join("cribo.toml");
            Self::try_load_and_combine(&mut config, &user_config_path, "user config")?;
        }

        // 3. Load project config (cribo.toml in current directory)
        let project_config_path = PathBuf::from("cribo.toml");
        Self::try_load_and_combine(&mut config, &project_config_path, "project config")?;

        // 4. Apply environment variables
        let env_config = EnvConfig::from_env();
        config = env_config.apply_to(config);

        // 5. Load CLI-provided config (highest precedence)
        if let Some(cli_config_path) = cli_config_path {
            Self::try_load_and_combine(&mut config, cli_config_path, "CLI config")?;
        }

        // Final validation
        config.python_version().with_context(|| {
            format!(
                "Invalid target-version in final config: {}",
                config.target_version
            )
        })?;

        Ok(config)
    }
}
