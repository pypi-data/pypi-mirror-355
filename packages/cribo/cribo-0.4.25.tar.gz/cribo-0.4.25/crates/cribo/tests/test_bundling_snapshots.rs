#![allow(clippy::disallowed_methods)] // insta macros use unwrap internally

use std::fs;
use std::io::Write;
use std::path::Path;
use std::process::Command;
use tempfile::TempDir;

use cribo::config::Config;
use cribo::orchestrator::BundleOrchestrator;
use cribo::util::get_python_executable;
use pretty_assertions::assert_eq;

// Ruff linting integration for cross-validation
use ruff_linter::linter::{ParseSource, lint_only};
use ruff_linter::registry::Rule;
use ruff_linter::settings::{LinterSettings, flags};
use ruff_linter::source_kind::SourceKind;
use ruff_python_ast::PySourceType;

/// Structured execution results for better snapshot formatting
#[derive(Debug)]
#[allow(dead_code)] // Fields are used via Debug trait for snapshots
struct ExecutionResults {
    status: ExecutionStatus,
    stdout: String,
    stderr: String,
}

/// Get filters for normalizing paths and Python version differences in snapshots
fn get_path_filters() -> Vec<(&'static str, &'static str)> {
    vec![
        // Python installation paths (minimal filtering needed with 5-line stderr limit)
        // macOS Homebrew Python paths
        (
            r"/opt/homebrew/Cellar/python@[\d.]+/[\d._]+/Frameworks/Python\.framework/Versions/[\d.]+/lib/python[\d.]+/",
            "<PYTHON_LIB>/",
        ),
        // Unix system Python paths
        (r"/usr/lib/python[\d.]+/", "<PYTHON_LIB>/"),
        // Windows Python paths
        (r"C:\\Python\d+\\lib\\", "<PYTHON_LIB>/"),
        // Windows hosted tool cache paths (GitHub Actions)
        (
            r"C:\\hostedtoolcache\\windows\\Python\\[\d.]+\\x64\\Lib\\",
            "<PYTHON_LIB>/",
        ),
        // Replace line numbers that may vary between Python versions
        (
            r"line \d+, in import_module",
            "line <LINE>, in import_module",
        ),
        // Note: Only keeping first 2 lines of stderr eliminates most cross-platform differences
        // Note: File paths eliminated by using stdin execution (shows as <stdin>)
    ]
}

#[derive(Debug)]
#[allow(dead_code)] // Fields are used via Debug trait for snapshots
enum ExecutionStatus {
    Success,
    Failed(i32),
}

/// Ruff linting results for cross-validation
#[derive(Debug)]
#[allow(dead_code)] // Fields are used via Debug trait for snapshots
struct RuffLintResults {
    f401_violations: Vec<String>, // Unused imports
    f404_violations: Vec<String>, // Late future imports
    other_violations: Vec<String>,
    total_violations: usize,
}

/// Run ruff linting on bundled code to cross-validate import handling
fn run_ruff_lint_on_bundle(bundled_code: &str) -> RuffLintResults {
    // Create settings for multiple import-related rules with both F401 and F404 enabled
    let settings = LinterSettings {
        rules: [Rule::UnusedImport, Rule::LateFutureImport]
            .into_iter()
            .collect(),
        ..LinterSettings::default()
    };

    let path = Path::new("<bundled>.py");
    let source_kind = SourceKind::Python(bundled_code.to_string());

    let result = lint_only(
        path,
        None,
        &settings,
        flags::Noqa::Enabled,
        &source_kind,
        PySourceType::Python,
        ParseSource::None,
    );

    let mut f401_violations = Vec::new();
    let mut f404_violations = Vec::new();
    let mut other_violations = Vec::new();

    for message in &result.messages {
        if let Some(rule) = message.to_rule() {
            let location = message.compute_start_location();
            let violation_info = format!(
                "Line {}: {} - {}",
                location.line.get(),
                rule.noqa_code(),
                message.body()
            );

            match rule {
                Rule::UnusedImport => f401_violations.push(violation_info),
                Rule::LateFutureImport => f404_violations.push(violation_info),
                _ => other_violations.push(violation_info),
            }
        }
    }

    let total_violations = f401_violations.len() + f404_violations.len() + other_violations.len();

    RuffLintResults {
        f401_violations,
        f404_violations,
        other_violations,
        total_violations,
    }
}

/// Test bundling fixtures using Insta's glob feature
/// This discovers and tests all fixtures automatically
#[test]
fn test_bundling_fixtures() {
    insta::glob!("fixtures/bundling", "*/main.py", |path| {
        // Extract fixture name from the path
        let fixture_dir = path.parent().unwrap();
        let fixture_name = fixture_dir.file_name().unwrap().to_str().unwrap();

        // Print which fixture we're running (will only show when not filtered out)
        eprintln!("Running fixture: {}", fixture_name);

        // Check if this is an expected failure fixture
        let expects_failure = fixture_name.starts_with("xfail_");

        // First, run the original fixture to ensure it's valid Python code
        let python_cmd = get_python_executable();
        let original_output = Command::new(&python_cmd)
            .arg(path)
            .current_dir(fixture_dir)
            .stdout(std::process::Stdio::piped())
            .stderr(std::process::Stdio::piped())
            .spawn()
            .and_then(|child| child.wait_with_output())
            .expect("Failed to execute original fixture");

        // If the original fixture doesn't run correctly, skip bundling
        if !original_output.status.success() && !expects_failure {
            let stderr = String::from_utf8_lossy(&original_output.stderr);
            let stdout = String::from_utf8_lossy(&original_output.stdout);

            panic!(
                "Original fixture '{}' failed to execute:\n\
                Exit code: {}\n\
                Stdout:\n{}\n\
                Stderr:\n{}\n\n\
                Fix the fixture before testing bundling.",
                fixture_name,
                original_output.status.code().unwrap_or(-1),
                stdout.trim(),
                stderr.trim()
            );
        }

        // Store original execution results for comparison
        let original_stdout = String::from_utf8_lossy(&original_output.stdout)
            .trim()
            .replace("\r\n", "\n")
            .to_string();
        let original_exit_code = original_output.status.code().unwrap_or(-1);

        // Create temporary directory for output
        let temp_dir = TempDir::new().unwrap();
        let bundle_path = temp_dir.path().join("bundled.py");

        // Configure bundler
        let config = Config::default();
        let mut bundler = BundleOrchestrator::new(config);

        // Bundle the fixture
        bundler.bundle(path, &bundle_path, false).unwrap();

        // Read the bundled code
        let bundled_code = fs::read_to_string(&bundle_path).unwrap();

        // Optionally validate Python syntax before execution
        let python_cmd = get_python_executable();
        let syntax_check = Command::new(&python_cmd)
            .args(["-m", "py_compile", "-"])
            .stdin(std::process::Stdio::piped())
            .stdout(std::process::Stdio::piped())
            .stderr(std::process::Stdio::piped())
            .spawn();

        if let Ok(mut child) = syntax_check {
            if let Some(mut stdin) = child.stdin.take() {
                let _ = stdin.write_all(bundled_code.as_bytes());
            }
            if let Ok(output) = child.wait_with_output() {
                if !output.status.success() && std::env::var("RUST_TEST_VERBOSE").is_ok() {
                    eprintln!(
                        "Warning: Bundled code has syntax errors for fixture {}",
                        fixture_name
                    );
                    eprintln!("Stderr: {}", String::from_utf8_lossy(&output.stderr));
                }
            }
        }

        // Run ruff linting for cross-validation
        let ruff_results = run_ruff_lint_on_bundle(&bundled_code);

        // Execute the bundled code via stdin for consistent snapshots
        let python_output = Command::new(&python_cmd)
            .arg("-")
            .stdin(std::process::Stdio::piped())
            .stdout(std::process::Stdio::piped())
            .stderr(std::process::Stdio::piped())
            .current_dir(temp_dir.path())
            .spawn()
            .and_then(|mut child| {
                if let Some(mut stdin) = child.stdin.take() {
                    let _ = stdin.write_all(bundled_code.as_bytes());
                }
                child.wait_with_output()
            })
            .expect("Failed to execute Python");

        // Check for unexpected Python execution failures
        if !python_output.status.success() && !expects_failure {
            let stderr = String::from_utf8_lossy(&python_output.stderr);
            let stdout = String::from_utf8_lossy(&python_output.stdout);

            panic!(
                "Python execution failed unexpectedly for fixture '{}':\n\
                Exit code: {}\n\
                Stdout:\n{}\n\
                Stderr:\n{}\n\n\
                If this failure is expected, rename the fixture directory with 'xfail_' prefix.",
                fixture_name,
                python_output.status.code().unwrap_or(-1),
                stdout.trim(),
                stderr.trim()
            );
        }

        // Compare bundled execution to original execution
        let bundled_stdout = String::from_utf8_lossy(&python_output.stdout)
            .trim()
            .replace("\r\n", "\n")
            .to_string();
        let bundled_exit_code = python_output.status.code().unwrap_or(-1);

        // For non-xfail tests, stdout should match exactly
        if !expects_failure {
            assert_eq!(
                original_stdout, bundled_stdout,
                "\nBundled output differs from original for fixture '{}'",
                fixture_name
            );
        }

        // Exit codes should also match for non-xfail tests
        if !expects_failure {
            assert_eq!(
                original_exit_code, bundled_exit_code,
                "\nExit code differs for fixture '{}'",
                fixture_name
            );
        }

        // Check for xfail tests that are now passing
        if python_output.status.success() && expects_failure {
            let stdout = String::from_utf8_lossy(&python_output.stdout);
            let stderr = String::from_utf8_lossy(&python_output.stderr);

            panic!(
                "Expected fixture '{}' to fail (marked with xfail_), but it succeeded!\n\
                Exit code: 0\n\
                Stdout:\n{}\n\
                Stderr:\n{}\n\n\
                This test is now passing. Please remove the 'xfail_' prefix from the fixture directory name.",
                fixture_name,
                stdout.trim(),
                stderr.trim()
            );
        }

        // Create structured execution results
        let execution_status = if python_output.status.success() {
            ExecutionStatus::Success
        } else {
            ExecutionStatus::Failed(python_output.status.code().unwrap_or(-1))
        };

        let execution_results = ExecutionResults {
            status: execution_status,
            stdout: String::from_utf8_lossy(&python_output.stdout)
                .trim()
                .replace("\r\n", "\n")
                .to_string(),
            stderr: {
                let full_stderr = String::from_utf8_lossy(&python_output.stderr)
                    .trim()
                    .replace("\r\n", "\n");
                // Keep only first 2 lines to avoid cross-platform traceback differences
                full_stderr.lines().take(2).collect::<Vec<_>>().join("\n")
            },
        };

        // Use Insta's with_settings for better snapshot organization
        insta::with_settings!({
            snapshot_suffix => fixture_name,
            omit_expression => true,
            prepend_module_to_snapshot => false,
            filters => get_path_filters(),
        }, {
            // Snapshot the bundled code
            insta::assert_snapshot!("bundled_code", bundled_code);

            // Snapshot execution results with filters applied
            insta::assert_debug_snapshot!("execution_results", execution_results);

            // Snapshot ruff linting results
            insta::assert_debug_snapshot!("ruff_lint_results", ruff_results);
        });
    });
}
