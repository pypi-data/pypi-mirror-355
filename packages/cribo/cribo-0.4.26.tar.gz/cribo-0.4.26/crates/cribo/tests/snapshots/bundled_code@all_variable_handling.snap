---
source: crates/cribo/tests/test_bundling_snapshots.rs
input_file: crates/cribo/tests/fixtures/bundling/all_variable_handling/main.py
---
#!/usr/bin/env python3
# Generated by Cribo - Python Source Bundler
# https://github.com/ophidiarium/cribo

import sys
import types
message_1 = "from conflict_module"
SHARED_NAME = "conflict_module_version"
def helper_func():
    """Helper function used by other modules"""
    return "helper result"
def another_helper():
    """Another helper function"""
    return "another helper"
UTILS_CONSTANT = "utils value"
def __cribo_init___cribo_11029d_nested_package_submodule():
    if '__cribo_11029d_nested_package_submodule' in sys.modules:
        return sys.modules['__cribo_11029d_nested_package_submodule']
    module = types.ModuleType('__cribo_11029d_nested_package_submodule')
    module.__file__ = __file__ if '__file__' in globals() else None
    sys.modules['__cribo_11029d_nested_package_submodule'] = module
    sys.modules['nested_package.submodule'] = module
    __all__ = ["sub_function", "SUB_CONSTANT"]

    def sub_function():
        """Function from submodule"""
        return "Hello from submodule"
    module.sub_function = sub_function

    def _private_sub_func():
        """Private function in submodule"""
        return "private submodule function"
    SUB_CONSTANT = "submodule_value"
    module.SUB_CONSTANT = SUB_CONSTANT
    message = "from submodule"
    module.__all__ = ['sub_function', 'SUB_CONSTANT']
    return module
def __cribo_init___cribo_c56070_nested_package():
    if '__cribo_c56070_nested_package' in sys.modules:
        return sys.modules['__cribo_c56070_nested_package']
    module = types.ModuleType('__cribo_c56070_nested_package')
    module.__file__ = __file__ if '__file__' in globals() else None
    sys.modules['__cribo_c56070_nested_package'] = module
    sys.modules['nested_package'] = module
    sub_function = sys.modules['nested_package.submodule'].sub_function
    module.sub_function = sub_function
    module.helper_func = helper_func
    __all__ = ["exported_from_init", "sub_function"]

    def exported_from_init():
        """Function exported from package __init__.py"""
        return f"From init, using helper: {helper_func()}"
    module.exported_from_init = exported_from_init

    def _internal_init_func():
        """Internal function not exported"""
        return "internal"
    PACKAGE_CONSTANT = "from_package"
    module.__all__ = ['exported_from_init', 'sub_function']
    return module
def __cribo_init___cribo_8010fb_simple_module():
    if '__cribo_8010fb_simple_module' in sys.modules:
        return sys.modules['__cribo_8010fb_simple_module']
    module = types.ModuleType('__cribo_8010fb_simple_module')
    module.__file__ = __file__ if '__file__' in globals() else None
    sys.modules['__cribo_8010fb_simple_module'] = module
    sys.modules['simple_module'] = module
    __all__ = ["public_func", "CONSTANT"]

    def public_func():
        """A public function that should be exported."""
        return "Hello from public_func"
    module.public_func = public_func

    def _private_func():
        """A private function that should not be exported."""
        return "This is private"
    CONSTANT = 42
    module.CONSTANT = CONSTANT
    _PRIVATE_CONSTANT = "secret"

    class InternalClass:
        pass
    module.__all__ = ['public_func', 'CONSTANT']
    return module
__cribo_modules = {'nested_package.submodule': '__cribo_11029d_nested_package_submodule', 'nested_package': '__cribo_c56070_nested_package', 'simple_module': '__cribo_8010fb_simple_module'}
__cribo_init_functions = {'__cribo_11029d_nested_package_submodule': __cribo_init___cribo_11029d_nested_package_submodule, '__cribo_c56070_nested_package': __cribo_init___cribo_c56070_nested_package, '__cribo_8010fb_simple_module': __cribo_init___cribo_8010fb_simple_module}
class CriboBundledFinder:

    def __init__(self, module_registry, init_functions):
        self.module_registry = module_registry
        self.init_functions = init_functions

    def find_spec(self, fullname, path, target=None):
        if fullname in self.module_registry:
            synthetic_name = self.module_registry[fullname]
            if synthetic_name not in sys.modules:
                init_func = self.init_functions.get(synthetic_name)
                if init_func:
                    init_func()
            import importlib.util
            return importlib.util.find_spec(synthetic_name)
        return None
sys.meta_path.insert(0, CriboBundledFinder(__cribo_modules, __cribo_init_functions))
__cribo_init___cribo_8010fb_simple_module()
__cribo_init___cribo_11029d_nested_package_submodule()
__cribo_init___cribo_c56070_nested_package()
public_func = sys.modules['simple_module'].public_func
CONSTANT = sys.modules['simple_module'].CONSTANT
exported_from_init = sys.modules['nested_package'].exported_from_init
sub_function = sys.modules['nested_package.submodule'].sub_function
message = message_1
print("Testing simple module exports:")
print(f"public_func() = {public_func()}")
print(f"CONSTANT = {CONSTANT}")
print("\nTesting nested package exports:")
print(f"exported_from_init() = {exported_from_init()}")
print(f"sub_function() = {sub_function()}")
print("\nTesting conflict resolution:")
print(f"message = {message}")
simple_module = sys.modules['simple_module']
print(f"\nsimple_module.__all__ = {simple_module.__all__}")
nested_package = sys.modules['nested_package']
sub = sys.modules['nested_package.submodule']
print(f"submodule.__all__ = {sub.__all__}")
